### PR TITLE
[WIP] TimeSeries Feature 

### PR DESCRIPTION
**Pull request by mop9047**
This work in progress feature makes it possible to train time series data in your own neural network, along with the new folder in src called TimeSeries, there are also 4 new examples

### CODE REVIEW

**Comment by shiffman** [comment]
Wow, very excited to try out these examples and learn more about this work in our meeting!

File: examples/timeSeries-hand-gestures/sketch.js [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+  // setup video capture
+  video = createCapture(VIDEO);
+  video.size(640, 480);
+  video.hide();
+
+  // place UI elements
+  UI();
+
+  // set backend as either webgl or cpu
+  ml5.setBackend('webgl')
```
[the remark this answers is not in the document]
Comment (ziyuan-linn) [reply examples/timeSeries-hand-gestures/sketch.js:41]: Hi Gottfried, `await tf.ready();` should be called before any other tfjs-related calls. This will ensure that the asynchronously loaded `webgpu` backend is fully loaded before any other tfjs operation happens.

[the remark this answers is not in the document]
Comment (gohai) [reply examples/timeSeries-hand-gestures/sketch.js:41]: Thanks a ton @ziyuan-linn! @mop9047, could you check if adding this to the library code makes the example work without the `setBackend()` call in the example code?

File: examples/timeSeries-train-quickdraw/index.html [hunk 2]
```diff
@@ -0,0 +1,22 @@
+<!--
+  👋 Hello! This is an ml5.js example made and shared with ❤️.
+  Learn more about the ml5.js project: https://ml5js.org/
+  ml5.js license and Code of Conduct: https://github.com/ml5js/ml5-next-gen/blob/main/LICENSE.md
+ 
+  This example demonstrates How to train your own quickdraw classifier through ml5.TimeSeries.
```
[the remark this answers is not in the document]
Comment (mop9047) [reply examples/timeSeries-train-quickdraw/index.html]: Hi professor, yes I agree it was a little bit confusing, I changed the name into `mouse-gesture-classifier`, also, currently working on a utility to use RDP algorithm

File: examples/timeSeries-hand-gestures/index.html [hunk 3]
```diff
@@ -0,0 +1,22 @@
+<!--
+  👋 Hello! This is an ml5.js example made and shared with ❤️.
+  Learn more about the ml5.js project: https://ml5js.org/
+  ml5.js license and Code of Conduct: https://github.com/ml5js/ml5-next-gen/blob/main/LICENSE.md
+ 
+  This example demonstrates training a Sign Language classifier through ml5.TimeSeries.
```
[the remark this answers is not in the document]
Comment (mop9047) [reply examples/timeSeries-hand-gestures/index.html:6]: Hi professor, I agree, it might be too specific for casual users, thus I changed it into hand gesture classifier, at the end the training example you can choose whatever word you want, so its not gonna be an issue, however for the load model, its hard to think of a moving gesture that everyone knows (thumbs up comes to mind but movements are minimal) so I still continued with the ASL goodbye and hello and included some links so users can get to know how the actual gesture is done

File: examples/timeSeries-hand-gestures/sketch.js [hunk 4]
```diff
@@ -0,0 +1,217 @@
+/*
```
[the remark this answers is not in the document]
Comment (mop9047) [reply examples/timeSeries-hand-gestures/sketch.js:1]: I tried running this, I might have made a mistake but it changed other files too! not sure how to specify which files only to commit but I had to delete all the changes made on the other files

File: examples/timeSeries-hand-gestures/sketch.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mop9047) [reply examples/timeSeries-hand-gestures/sketch.js]: Hi @ziyuan-linn , thanks for the response, put the `tf.await() `in the init of the class and it worked well, however is the instantiation of the model in preload a necessity? @gohai I changed the examples as well, and now none of them uses `setBackend()`. not sure if the main branch is updated already but the examples in the NN class still has them, should we change them? thanks!

File: examples/timeSeries-hand-gestures/sketch.js [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    'I want to train: <br><br> 1.) Type any word you want to pair with a gesture, e.g. "HELLO" <br> 2.) Do the gesture associated to the word, make sure to do it until the points disappear. <br> 3.) Move your hand out of the frame and repeat the gesture, do this multiple times <br> 4.) Do the same for other words e.g. "BYE" <br> 5.) Once all data is collected, press Train and Save<br><br> Tip: have at least 5 datasets for each word'
+  );
+  instructionP.style("width", "640px");
+  dataCountsP = createP(
+    "-> After the gesture a tally will appear here <-"
+  );
+
+  train_but = createButton('Train and Save');
+  train_but.mouseClicked(trainModelAndSave);
+  train_but.style("font-family", "Georgia");
```
[the remark this answers is not in the document]
Comment (mop9047) [reply examples/timeSeries-hand-gestures/sketch.js:163]: Hi professor, tried to put the button and initial styles in the HTML file, although the length was not drastically changed, mainly because of the highlighting logic when a button is active, do you think I should remove this feature in favor of code simplicity? thanks!

File: examples/timeSeries-stock-prediction/index.html [hunk 6]
```diff
@@ -0,0 +1,22 @@
+<!--
+  👋 Hello! This is an ml5.js example made and shared with ❤️.
+  Learn more about the ml5.js project: https://ml5js.org/
+  ml5.js license and Code of Conduct: https://github.com/ml5js/ml5-next-gen/blob/main/LICENSE.md
+ 
+  This example demonstrates Training a Stock Price Predictor through ml5.TimeSeries.
```
[the remark this answers is not in the document]
Comment (mop9047) [reply examples/timeSeries-stock-prediction/index.html]: Hi professor, changed the example to weather predictor as you said, it works quite well and the dataset I put makes for dynamic predictions, however I am not sure how to visualize it, currently just making a table of values

File: examples/timeSeries-stock-prediction/sketch.js [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+function draw() {
+  // draw some helpful visualizations
+  updatedUI()
+}
+
+// create custom line graph for stock close prices
+let py = 300;
+let px = 80;
+function updatedUI(){
```
[the remark this answers is not in the document]
Comment (mop9047) [reply examples/timeSeries-stock-prediction/sketch.js]: deleted this file :)

File: examples/timeSeries-train-quickdraw/sketch.js [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+// code to signify drawing can be done again
+function mouseReleased(){
+  pressedOnce = true;
+}
+
+////////////// UI Elements ////////////
+
+// code to visualize how much ink left
+function inkBar(){
+  datapoints = map(frame_count,0,ink_multiplier*num_seq, 0,num_seq)
```
[the remark this answers is not in the document]
Comment (mop9047) [reply examples/timeSeries-train-quickdraw/sketch.js]: Hi professor, I am almost done with the RDP algorithm, and it works well, the code I referenced originally uses vectors, I am currently continuing this idea but do you think it might cause issues with the user's side? (i.e. they need to know use vectors only to make this work)

File: src/TimeSeries/index.js [hunk 9]
```diff
@@ -0,0 +1,743 @@
...
+   */
+
+  /* adding data: can only accept the following formats:
+     - for xInputs:
+      1. Sequence of objects (array of objects) 
+        [{x: , y: },{x: , y: },{x: , y: },{x: , y: }]
+      2. Sequence of arrays (array of array, order matters)
+        [[],[],[],[]]
+      3. Sequence of values (inputlabels should be provided by user)
+        [[,,,,,]] e.g. shape = {inputLabels: ['x','y']} will become [{x: , y: },{x: , y: },{x: , y: },{x: , y: }]
```
[the remark this answers is not in the document]
Comment (mop9047) [reply src/TimeSeries/index.js:80]: Hi professor, I included this format mainly to account for as many formats as possible, in this case the outer [] is not seen by the user and in their perspective they only sort of push all their values, regardless of sequence, inside one array only. (I think its the simplest as the user doesn't have to "group" sequences together) Anyways, tell me what you think and we can discuss if we should scrap it or not.

File: src/TimeSeries/index.js [hunk 10]
```diff
@@ -0,0 +1,743 @@
...
+  
+    - for yInputs:
+      1. similar to neural network, so use same logic
+
+    - at the end of the adding data, the data is formatted to a sequence of objects similar to 1 of xinputs
+ 
+    - new parameter dataModality, either spatial or sequential, spatial uses cnn1d and sequential uses lstm
```
[the remark this answers is not in the document]
Comment (mop9047) [reply src/TimeSeries/index.js]: Hi professor, I agree with this comment, I think that the dataModality is a little bit confusing, especially when the inputs are strings! as much as I want the data to be automatically categorized, unfortunately I currently cannot see a practical way to know, as other data such as weather also uses numbers exclusively, so instead of this, I changed data modality into `spatialData` which accepts a boolean, by default without setting it, it is using lstm which works for all data sequences so its not a problem, so if users want more accuracy for less datasets in coordinate data they can just set this as true, I think this is the most pragmatic way for now.

File: examples/timeSeries-train-quickdraw/index.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (gohai) [reply examples/timeSeries-train-quickdraw/index.html]: 👍  (Please call me just Gottfried : )

File: examples/timeSeries-hand-gestures/index.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (gohai) [reply examples/timeSeries-hand-gestures/index.html:6]: This seems like a very sensible compromise!

File: examples/timeSeries-hand-gestures/sketch.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (gohai) [reply examples/timeSeries-hand-gestures/sketch.js:1]: It's probably easiest to run the prettifier, and then selectively determine which changes to include. (In GitHub Desktop, this can be done via the context menu as you are exploring changes. You can accept - or revert changes there.)

File: examples/timeSeries-hand-gestures/sketch.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (gohai) [reply examples/timeSeries-hand-gestures/sketch.js]: Good catch, let me open a separate issue about this @mop9047.

File: examples/timeSeries-hand-gestures/sketch.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (gohai) [reply examples/timeSeries-hand-gestures/sketch.js:163]: IYou're right that ultimately in total lines of code the benefit might be negligible - but I'd still try to keep those concerns separated, and do styling in HTML/CSS, and merely the event handlers in JS.

File: examples/timeSeries-stock-prediction/index.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (gohai) [reply examples/timeSeries-stock-prediction/index.html]: Have you considered plotting them along the x axis?

File: examples/timeSeries-train-quickdraw/sketch.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (gohai) [reply examples/timeSeries-train-quickdraw/sketch.js]: The vector math is merely inside the RDP function? Then I think this shouldn't be all too confusing

File: src/TimeSeries/index.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (gohai) [reply src/TimeSeries/index.js:80]: What are examples for the different cases? (Maybe easiest to look at a how this would function from the user side)

File: src/TimeSeries/index.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (gohai) [reply src/TimeSeries/index.js]: Still thinking that having different `tasks` might be more clear and educational compared to a flag (where we don't quite know what its doing without looking at the source code).

File: examples/timeSeries-hand-gestures/sketch.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mop9047) [reply examples/timeSeries-hand-gestures/sketch.js]: @gohai , I found an issue with this implementation which it actually only works in safari, otherwise I am getting a `WebGPU readSync is only available for CPU-resident tensors` error in chrome and safari, not sure how to fix this, but what I did instead was put the `setBackend()` line inside of the time series class and used `webgl` not sure if this is a good fix, but it fixed the error for running in other browsers such as operagx and chrome

File: src/TimeSeries/index.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (mop9047) [reply src/TimeSeries/index.js]: Hi Gottfried, I reverted to the old way without the flag. I implemented a change for this and changed the `dataModality` into `dataMode` to make it shorter. I keep the `spatial` but changed the `sequential` to linear, tell me what you think about these changes

File: examples/timeSeries-train-quickdraw/sketch.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (mop9047) [reply examples/timeSeries-train-quickdraw/sketch.js]: Yes you're right! I just finished implementing the RDP and its a timeSeries utility called `padCoordinates(seq_to_pad, target_seq_size)`, I specifically chose this name to introduce the idea of padding and its importance in dealing with time series data. The example is in the mouse Gesture RDP folder. 

That said should I remove the ink one entirely? thanks!

File: examples/timeSeries-stock-prediction/index.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (mop9047) [reply examples/timeSeries-stock-prediction/index.html]: I could plot it but it was the exact problem I had with the stock predictor where the code was getting complicated just because of the UI, that said, I implemented a simple dynamic table that shows only the latest values, so it still has a responsive feel to it. you can check it out in the example if you'd like!

File: examples/timeSeries-hand-gestures/sketch.js [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+  model.save('model');
+}
+
+function gotResults(results){
+  predicted_word = results[0].label
+  console.log(predicted_word)
+  text(predicted_word, 200,200)
+}
+
+function UI(){
```
[the remark this answers is not in the document]
Comment (mop9047) [reply examples/timeSeries-hand-gestures/sketch.js]: Hi professor, I did this for both the mouse gesture and the weather predictor, they now only have buttons but have helpful texts when pressed. For the hand gestures, specifically for the train and save, I think there is a way to sort of minimize the text at the bottom, however, I patterned the UI from the train and save of the Neural network and it might be helpful if there is a consistency between the two. I am open to both and am currently drafting code for alternate UI

File: src/TimeSeries/index.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (mop9047) [reply src/TimeSeries/index.js:72]: Hi Gottfried, I decided that this format is not actually useful as the batch size needs to be specified as well, which leads to some nasty complications when user input is not right. that said I will probably just remove this and stick with the first two.
